Timeline: More editable control

simo9000 opened the pull request:
I resisted the urge to do this but I didn't see a way around it. #2284 provided a clear means of decided when level of options was in control of the editable behavior of items. However it fell short in that when individual items are determined to be editable (and `overrideItems` is false), it enabled all three edit behaviors. This enables further control over individual item behavior but allowing `updateTime`, `updateGroup`, and `remove` to be toggled at the item level. I am glad I chose to pursue this as I found a bug which was introduced in #2284 where `timeline.editable == true` would also set `overrideItems` to true.

yotamberk commented on lib/timeline/component/ItemSet.js:
```diff
@@ -1388,12 +1388,14 @@ ItemSet.prototype._onDrag = function (event) {
 
-      var updateTimeAllowed = me.options.editable.updateTime || (
-          props.item.editable === true && !me.options.editable.overrideItems);
+      var updateTimeAllowed = (me.options.editable.overrideItems && me.options.editable.updateTime) || (
```
This overrides the previous default... what happens when:
`me.options.editable.overrideItems==false` and `me.options.editable.updateTime==true`?
updating time in this case should be allowed even if `props.item.editable.updateTime` is not defined (or false, as set in default)

yotamberk commented on lib/timeline/component/ItemSet.js:
```diff
@@ -1456,8 +1458,8 @@ ItemSet.prototype._onDrag = function (event) {
-      var updateGroupAllowed = me.options.editable.updateGroup ||
-          props.item.editable === true;
+      var updateGroupAllowed = (me.options.editable.overrideItems && me.options.editable.updateGroup) ||
+          props.item.editable.updateGroup;
```
What about: `!me.options.editable.overrideItems && me.options.editable.updateGroup`? You should be able to update groups but in this logic, you can't...

yotamberk commented on lib/timeline/component/item/Item.js:
```diff
@@ -148,7 +164,7 @@ Item.prototype.repositionY = function() {
                   this.options.editable.overrideItems)
-                  || (this.data.editable === true &&
+                  || (this.editable.remove &&
                   !this.options.editable.overrideItems);
```
What about: `this.data.editable.remove && !this.options.editable.overrideItems`?

simo9000 commented on lib/timeline/component/ItemSet.js:
```diff
@@ -1225,8 +1225,8 @@ ItemSet.prototype._onDragStart = function (event) {
     // override options.editable
-    if ((item.editable === false || !item.editable)
-    && !this.options.editable.overrideItems) {
+    if (!item.editable.updateTime && !item.editable.updateGroup
```
This predicate is in the `ItemSet._onDragStart`, my understanding is that  `remove` is only concerned with `Item._repaintDeleteButton`. Shouldn't this predicate not be affected by the value of `remove`?

yotamberk replied:
oops you're right. I didn't notice the function it was in. Ignore this comment

simo9000 commented:
@yotamberk Prior to this PR, there was no default for `item.editable`, the behavior was actually to enable editing if `item.editable` is not set as configured by the timeline. To keep this behavior, i have removed the defaults for `item.editable` to allow it to be null and updated the predicates. This oversight was actually made when i set up the [test matrix](https://docs.google.com/spreadsheets/d/1p3ldpknj067XwmCbv4sEPihogsXTx4qdGBpnTxyWunY/edit?usp=sharing) for #2284. I have changed the anticipated behavior of default for this commit. Please reconsider for approval.

Thanks for catching the logic flaws.

yotamberk commented on lib/timeline/component/ItemSet.js:
```diff
@@ -1456,8 +1459,8 @@ ItemSet.prototype._onDrag = function (event) {
 
-      var updateGroupAllowed = me.options.editable.updateGroup ||
-          props.item.editable === true;
+      var updateGroupAllowed = ((me.options.editable.overrideItems || !props.item.editable ) && me.options.editable.updateGroup) || (
```
Here too:
`(me.options.editable.overrideItems && me.options.editable.updateGroup) || 
(!me.options.editable.overrideItems && (me.options.editable==true || me.options.editable.updateGroup)`

yotamberk commented:
I hope I'm not discouraging you. The logic is kind of messy and I think it needs to be cleaned up a bit before being approved

simo9000 commented:
@yotamberk It's not discouraging at all. You are right that it was still messy. While it was "working" to pass all the tests I could come up with, the predicates were not standardized even though they were doing the same basic tests. I decided to formalize this logic:

`Ov = options.editable.overrideItems`
`<property> => updateGroup || updateTime || remove`
`Tp = options.editable.<property>`
`I = item.editable`
`Ip = item.editable.<property>`

The abstract predicate that applies in all cases is 

`((Ov || I == null) && Tp) || (!Ov && I != null && Ip)`

I have changed all three predicates you suggested changes on to use this form. The predicate at line 1460 of itemset.js is actually redundant as `updateGroupAllowed` is still in scope from the predicate at line 1364 so i have removed it.

Your code suggestions included `item.editable == true`, this will not work with the edits that i have made to Item.js. Item.editable is only assigned values of null or of object. As the abstract predicate indicates, i have removed the reliance on truthiness.

yotamberk commented:
MUCH BETTER! Good job! 
>Item.editable is only assigned values of null or of object.

I'm still wondering about that. seems a little weird that it's different than `options.editable`. But I'll approve this. Again, very nice work! Looking forward to more PRs =)

mojoaxel commented:
@simo9000 Thanks! Thats a nice improvement!
